### ISSUE TITLE
Feature Request: `curly` rule should optionally ignore unnecessary braces

### ISSUE THREAD

**Issue by ben-page** [eslint/eslint#6141]
**What version of ESLint are you using?**
v2.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

I think the `curly` rule should have an option to relax complaints about unnecessary braces. Many style guides have rules to enforcing the use of curly braces, but few of them enforce the removal of braces. It's often desirable to allow the programmer to add additional braces at his/her discretion.

**Comment by kaicataldo** [comment]
Thanks for the issue - looks like you already created a PR for it. We normally recommend waiting until the suggested enhancement is approved before working on an issue, so as to avoid unnecessary/extra work. Let's see what everyone has to say, and thanks for the contribution, regardless!

**Comment by ilyavolodin** [comment]
@ben-page Could you provide example of the code that would currently cause an issue, and your configuration?

**Comment by BYK** [comment]
@ben-page I'd love to see a full proposal with code examples and why existing options are not enough for those cases before moving forward with this.

**Comment by ben-page** [comment]
A example how of the `curly` rule in ESLint works now:
```javascript
/*eslint curly: ["error", "multi-or-nest"]*/

if (!foo) //3:1  error  Expected { after 'if' condition      curly
    bar = {
        foo: 'bar'
    };

if (foo) { //8:1  error  Unnecessary { after 'if' condition   curly
    doSomething();
} else { //10:3  error  Unnecessary { after 'else'           curly
    doSomethingElse();
}
```

Personally, I want to enforce the rule on line 3. But I don't really care about the warnings on lines 8 and 10. I'd like to allow the programmer that much discretion. Now imagine applying ESLint to a existing large project. The second and third warnings, increase the amount of work to bring existing code in line with the style, but they don't improve the code much.

So, I'm proposing an option to suppress the "Unnecessary {" warnings. It could look like this:

```javascript
/*eslint curly: ["error", "multi-or-nest", {allowExtra: true}]*/

if (!foo) //3:1  error  Expected { after 'if' condition      curly
    bar = {
        foo: 'bar'
    };

if (foo) {
    doSomething();
} else {
    doSomethingElse();
}
```

**Comment by ben-page** [comment]
@BYK What is missing to qualify as a full proposal? I think I included everything described in [Change Requests](http://eslint.org/docs/developer-guide/contributing/changes).

**Comment by BYK** [comment]
@ben-page you just did with https://github.com/eslint/eslint/issues/6141#issuecomment-218882841 :)

**Comment by BYK** [comment]
I'm not a strong :+1: but it seems like a fair request to me. I personally prefer stronger consistency but from an objective standpoint, this looks like a possible use case. @eslint/eslint-team what do you think?

**Comment by ilyavolodin** [comment]
I'm still confused. Why do you want to require braces on line 3 but now on line 8? Is it because of multi-line statement?

**Comment by ben-page** [comment]
I want warnings about missing braces, but about not unnecessary braces. If someone feels extra braces makes things more readable, I'm fine with that. But I don't want them deciding braces are unnecessary.

**Comment by mikesherov** [comment]
I'm also confused... What about line 3 makes it optional? Multiple lines?

**Comment by ben-page** [comment]
3 isn't optional 8 and 10 are. All I want is messages about removing braces to be optional.

**Comment by ilyavolodin** [comment]
@ben-page OK, so if I understand it correctly, you want to maintain the same behavior as currently done by `multi-or-nested` option, but not warn when braces are optional? We've had requests for tri-state behavior in the past for other rules (required, disallow and don't care), and in every case consensus was that if you don't care, you should turn the rule off with comments.

**Comment by BYK** [comment]
@ilyavolodin I don't think turning the rule off using comments for every single brace is a practical solution. The user would simply remove the brave instead which defeats the purpose.

**Comment by nzakas** [comment]
Here's what I understand this proposal to be:

1. The rule should warn when curlies are required but are missing.
2. The rule should not warn when curlies are present but are not required. 

Assuming that's the correct description of the proposal, please indicate your support by adding a 👍 to this comment.

**Comment by ben-page** [comment]
@nzakas, yes exactly. Thank you.👍

**Comment by nzakas** [comment]
@ilyavolodin it looks like you objected to the proposal. Can you please explain so we can finish the discussion?

**Comment by ilyavolodin** [comment]
As far as I understand the proposal, this would require tri-state option. We always tried to avoid this. If this can be done with a boolean option, I'm fine with the change.

**Comment by nzakas** [comment]
Hmm, I'm not seeing tri-state behavior here. The proposal is to add an option that suppresses all "unnecessary" warnings.

**Comment by ilyavolodin** [comment]
I'm simplifying it a little bit, but right now we have two options, always require curly braces, and disallow them if it's a single statement. This adds a third state "don't care". Now, our options are not laid out that way, so maybe my concerns have no base, but when you oversimplify what this rule does, we are adding third state here, which we always tried to avoid.
Please correct me if I'm wrong.

**Comment by ilyavolodin** [comment]
OK, I'll withdraw my objection, since this is not `true`, `false` case. Looks like the issue can be accepted now.

**Comment by dcleao** [comment]
Just what I was looking for...

**Comment by kaicataldo** [comment]
Closing this as there hasn't been enough interest to push this forward in quite a while. Thanks for contributing!

**Comment by dcleao** [comment]
@kaicataldo Really? Just because people are "tired" of waiting... it doesn't mean they're not interested :-( Also, new people come here, see how long its been, and don't even comment...

**Comment by kaicataldo** [comment]
We've found that issues that get buried in the backlog without getting accepted end up getting lost (as can be evidenced by the lack of discussion here). Keeping our issues backlog manageable is necessary for keeping the project moving forward, both for us maintainers and also for our contributors.

Feel free to open a new issue with your proposal if you'd like to get some discussion going around this - chances are much higher it'll get some attention when it's not multiple pages deep into our issues list!

**Comment by dcleao** [comment]
@kaicataldo That's a great management practice.
